packages/next-swc/crates/next-core/src/next_import_map.rs

```diff
@@ -818,8 +825,56 @@ async fn insert_next_shared_aliases(
     import_map.insert_singleton_alias("@swc/helpers", get_next_package(project_path));
     import_map.insert_singleton_alias("styled-jsx", get_next_package(project_path));
     import_map.insert_singleton_alias("next", project_path);
-    import_map.insert_singleton_alias("react", project_path);
-    import_map.insert_singleton_alias("react-dom", project_path);
+
+    let ppr = *next_config.enable_ppr().await?;
+    let taint = *next_config.enable_taint().await?;
+    let react_channel = if ppr || taint { "-experimental" } else { "" };
+
+    let mut precompiled = |key, value: String| {
+        import_map.insert_exact_alias(key, request_to_import_mapping(project_path, &value));
+    };
+
+    precompiled("react", format!("next/dist/compiled/react{react_channel}"));
+    precompiled(
+        "react/jsx-runtime",
+        format!("next/dist/compiled/react{react_channel}/jsx-runtime"),
+    );
+    precompiled(
+        "react/jsx-dev-runtime",
+        format!("next/dist/compiled/react{react_channel}/jsx-dev-runtime"),
+    );
+    precompiled(
+        "react-dom",
+        format!("next/dist/compiled/react-dom{react_channel}"),
+    );
+    precompiled(
+        "react-dom/server",
+        format!("next/dist/compiled/react-dom{react_channel}/server"),
+    );
+    precompiled(
+        "react-dom/server.browser",
+        format!("next/dist/compiled/react-dom{react_channel}/server.browser"),
+    );
+    precompiled(
+        "react-dom/server.edge",
+        format!("next/dist/compiled/react-dom{react_channel}/server.edge"),
+    );
+    precompiled(
+        "react-dom/client",
+        format!("next/dist/compiled/react-dom{react_channel}/client"),
+    );
+    precompiled(
+        "react-dom/static",
+        format!("next/dist/compiled/react-dom-experimental/static"),
+    );
+    precompiled(
+        "react-dom/static.edge",
+        format!("next/dist/compiled/react-dom-experimental/static.edge"),
+    );
+    precompiled(
+        "react-dom/static.browser",
+        format!("next/dist/compiled/react-dom-experimental/static.browser"),
+    );
```

Comment:
a wildcard alias should work here?

Reply:
oh wait we already have the helper function below

Reply:
I fixed it by moving it to aliases for `client` + `pages`. (My bad; it was not in the app directory)
Also, I cleaned it up by using a wildcard alias.